### PR TITLE
phpunit9 and php8

### PR DESCRIPTION
**Pull request by delboy1978uk**
This cant be merged until the following are PHP 8 ready, as I have changed the repos/branches in the composer.json in order that we could work on this:
laminas/laminas-crypt#11
laminas/laminas-mime#14
laminas/laminas-math#5

|    Q          |   A
|-------------- | ------
| Documentation | /no
| Bugfix        | /no
| BC Break      | /no
| New Feature   | yes/
| RFC           | /no
| QA            | /no

### Description
get mail working on php8 and phpunit9

### CODE REVIEW

**Comment by delboy1978uk** [comment]
This cant be merged until the following are PHP 8 ready, as I have changed the repos/branches in the `composer.json` in order that we could work on this:
https://github.com/laminas/laminas-crypt/pull/11 
https://github.com/laminas/laminas-mime/pull/14 
https://github.com/laminas/laminas-math/pull/5

**Comment by glensc** [comment]
@delboy1978uk, my 0.2c:

1. you should rebase, not merge upstream changes. it's difficult to review if your PR commits 59 commits that are not even yours. `git push -f` to update branch after rebase (don't create another PR!)
2. you can mark PR as "draft", if it's not ready for merge (waiting for dependencies)
3. you can probably put the waiting for merge note in PR body, more visible than single comment at bottom of the PR after tons of changes (most of the noise is due merge commits included in PR).

File: composer.json [hunk 1]
```diff
@@ -24,22 +24,37 @@
             "config-provider": "Laminas\\Mail\\ConfigProvider"
         }
     },
+    "minimum-stability": "dev",
+    "repositories": [
+        {
+            "type": "git",
+            "url": "https://github.com/delboy1978uk/laminas-crypt.git"
```
Comment (Ocramius) [inline composer.json:31]: To be dropped before merge

File: composer.json [hunk 2]
```diff
@@ -24,22 +24,37 @@
...
+        },
+        {
+            "type": "git",
+            "url": "https://github.com/bfoosness/laminas-mime"
```
Comment (Ocramius) [inline composer.json:35]: To be dropped before merge

File: composer.json [hunk 3]
```diff
@@ -24,22 +24,37 @@
...
+        },
+        {
+            "type": "git",
+            "url": "https://github.com/bfoosness/laminas-math.git"
```
Comment (Ocramius) [inline composer.json:39]: To be dropped before merge

File: composer.json [hunk 4]
```diff
@@ -24,22 +24,37 @@
...
+        }
+    ],
     "require": {
-        "php": "^7.1",
+        "php": "^7.1 || ^8.0",
```
Comment (Ocramius) [inline composer.json]: `^7.3 || ^8.0` if we use PHPUnit 9.5+

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: composer.json (hunk 4 above)
Comment (delboy1978uk) [reply composer.json]: done

**Comment by delboy1978uk** [comment]
Hi @Ocramius , I'm a little confused as to why the build is failing on that last commit https://github.com/laminas/laminas-mail/pull/127/commits/e4f70edd1c9314de7fb51d77afae5a7cf9c1a130? Also it passes on my own travis page? https://travis-ci.com/github/delboy1978uk/laminas-mail/builds/213443818
Do I need to keep rebasing it periodically?

**Comment by Ocramius** [comment]
@delboy1978uk don't really understand the failure either:

```
1) Error

The data provider specified for LaminasTest\Mail\Header\ContentDispositionTest::testParameterWrapping is invalid.

PHPUnit\Framework\InvalidDataProviderException: The key "With two ordered items" has already been defined in the data provider "parameterWrappingProvider".
```

This seems to be related with a phpunit upgrade, perhaps.

The warning on opening a temporary file is more complex, but unsure if related to travis-ci infrastructure.

**Comment by glensc** [comment]
@Ocramius For https://github.com/laminas/laminas-mail/pull/127#issuecomment-763705043

I already replied via https://github.com/laminas/laminas-mail/pull/117#discussion_r559250424, it's fixed with this PR:
- https://github.com/laminas/laminas-mail/pull/129

why there are two PR's doing the same?

**Comment by delboy1978uk** [comment]
@glensc that might have been my fault, I see the other PR was there first now. I was loading one of my own packages into PHP8 and I noticed it still wasn't ready, so without a thought I just downloaded to help where I can. What's the vibe to get your data provider key PR into my branch, do I cherry pick it or just wait for the merges to happen?

**Comment by glensc** [comment]
I pinged @Ocramius to merge https://github.com/laminas/laminas-mail/pull/129#issuecomment-764825005, he's been active recently in this project. So I'd rebase once that','s merged, or if want to fix CI, cherry-pick and convert PR to Draft to prevent premature merge.

This project does not have a CONTRIBUTING.md guide, so I'm don't what's the procedures. If this ould be my project, I would ask first to rebase and squash some commits, maybe even extract changes to separate PR to merge ahead. usually, I've pointed people to these guides, how to make good commits and commit messages. 39 commits here seems way too much.
- https://www.freshconsulting.com/atomic-commits/
- https://chris.beams.io/posts/git-commit/

**Comment by froschdesign** [comment]
@glensc 
> This project does not have a CONTRIBUTING.md guide…

This project uses the GitHub functionality for this and therefore, under each text field you will find:

> Remember, contributions to this repository should follow its [contributing guidelines](https://github.com/laminas/.github/blob/b0a1d63a0dd2472e38a660fd89b23f443b4a7b3d/CONTRIBUTING.md) and [code of conduct](https://github.com/laminas/.github/blob/b0a1d63a0dd2472e38a660fd89b23f443b4a7b3d/CODE_OF_CONDUCT.md).

**Comment by Ocramius** [comment]
These changes are now included through #117, thanks @delboy1978uk!

**Pull request closed without merging**